src/stream/src/common/table/state_table_cache.rs

```diff
@@ -0,0 +1,477 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use risingwave_common::array::Op;
+use risingwave_common::estimate_size::EstimateSize;
+use risingwave_common::row::{OwnedRow, Row, RowExt};
+use risingwave_common::types::{DefaultOrdered, ScalarRefImpl};
+
+use crate::common::cache::{StateCache, TopNStateCache};
+
+/// The watermark cache key is just an `OwnedRow` wrapped in `DefaultOrdered`.
+/// This is because we want to use the `DefaultOrdered` implementation of `Ord`.
+/// The assumption is that the watermark column is the first column in the row.
+/// So it should automatically be ordered by the watermark column.
+/// We disregard the ordering of the remaining PK datums.
+///
+/// TODO(kwannoel):
+/// We can also store bytes from encoded pk.
+/// However, we need to ignore NULL values in their encoded form.
+/// Additionally, we need to encode pk w/o vnode, to make sure watermark stays as the prefix.
+/// The benefit is we avoid the clone of Arc of pk indices.
+type WatermarkCacheKey = DefaultOrdered<OwnedRow>;
+
+/// Cache Invariants
+/// -----------------
+/// - If cache is synced, it will ALWAYS contain TopN entries. NOTE: TopN here refers to the lowest
+///   N entries, where the first is the minimum and so on.
+/// - Cache will not contain NULL values, they are ignored in watermark state cleaning. They will
```

Comment:
(Maybe off-topic) How can there be NULL values? (You mean rows with `NULL`  watermark column, right?)

Reply:
Yes